<!--
  ~ The Exomiser - A tool to annotate and prioritize genomic variants
  ~
  ~ Copyright (c) 2016-2021 Queen Mary University of London.
  ~ Copyright (c) 2012-2016 Charité Universitätsmedizin Berlin and Genome Research Ltd.
  ~
  ~ This program is free software: you can redistribute it and/or modify
  ~ it under the terms of the GNU Affero General Public License as
  ~ published by the Free Software Foundation, either version 3 of the
  ~ License, or (at your option) any later version.
  ~
  ~ This program is distributed in the hope that it will be useful,
  ~ but WITHOUT ANY WARRANTY; without even the implied warranty of
  ~ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  ~ GNU Affero General Public License for more details.
  ~
  ~ You should have received a copy of the GNU Affero General Public License
  ~ along with this program.  If not, see <http://www.gnu.org/licenses/>.
  -->

<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml"
      xmlns:th="http://www.thymeleaf.org">

    <head>
        <title>Exomiser :: Results</title>
        <meta charset="UTF-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1.0"/>
        <link rel='icon' type='image/x-icon' href="favicon.ico" th:href="@{/favicon.ico}"/>
        <link href="https://cdn.jsdelivr.net/webjars/bootstrap/3.3.7-1/css/bootstrap.min.css" th:href="@{/webjars/bootstrap/3.3.7-1/css/bootstrap.min.css}" rel="stylesheet" media="screen"/>
    </head>
    <body>
    <nav class="navbar navbar-default">
        <div class="container">
            <div class="navbar-header">
                <button type="button" class="navbar-toggle collapsed" data-toggle="collapse"
                        data-target=".navbar-collapse">
                    <span class="sr-only">Toggle navigation</span>
                    <span class="icon-bar"></span>
                    <span class="icon-bar"></span>
                    <span class="icon-bar"></span>
                </button>
                <a class="navbar-brand" href="/" th:href="@{/}">
                    <img src="../static/logo.png" th:src="@{logo.png}" width="auto" height="30"/>
                </a>
            </div>
            <div class="navbar-collapse collapse">
                <ul class="nav navbar-nav">
                    <li><a href="index.html" th:href="@{index}">Home</a></li>
                    <li class="active"><a href="submit.html" th:href="@{submit}">Analyse</a></li>
                    <li><a href="publications.html" th:href="@{publications}">Publications</a></li>
                    <li><a href="download.html" th:href="@{download}">Download</a></li>
                    <li><a href="legal.html" th:href="@{legal}">Legal</a></li>
                    <li><a href="about.html" th:href="@{about}">About</a></li>
                </ul>
            </div><!--/.nav-collapse -->
        </div>
    </nav>
        <div id="content" class="container">
            <h2>
                Analysis Results for
                <th:block th:text="${sampleName}"> Slartibartfast</th:block>
            </h2>
            <ul class="nav nav-pills" id="tabs">
                <li><a href="#settings" role="tab" data-toggle="tab"><b><span class="glyphicon glyphicon-cog"></span> Analysis Settings</b></a></li>
                <li><a href="#filtering" role="tab" data-toggle="tab"><b><span class="glyphicon glyphicon-filter"></span>
                    Filtering Summary</b></a></li>
                <li><a href="#distribution" role="tab" data-toggle="tab"><b><span class="glyphicon glyphicon-align-left"></span> Variant Type Distribution</b></a></li>
                <li class="active"><a href="#prioritisation" role="tab" data-toggle="tab"><b><span class="glyphicon glyphicon-sort"></span> Prioritised Genes</b></a></li>
                <li th:unless="${#lists.isEmpty(unAnalysedVarEvals)}"><a href="#errors" role="tab" data-toggle="tab"><b><span class="glyphicon glyphicon-warning-sign"></span> Unanalysed Variants</b></a></li>
                <!--<li><a href="#about" role="tab" data-toggle="tab"><b><span class="glyphicon glyphicon-question-sign"/> About</b></a></li>-->
            </ul>
            <div class="tab-content">
                <div class="tab-pane fade" id="settings">
                    <p>Settings used for this analysis:</p>
                    <pre th:text="${settings}">ExomiserSettings{vcfFilePath=Pfeiffer.vcf, pedFilePath=null, prioritiser=exomiser-mouse, maximumFrequency=1.0, minimumQuality=0.0, geneticInterval=null, removePathFilterCutOff=false, removeDbSnp=false, removeOffTargetVariants=true, candidateGene=, modeOfInheritance=AUTOSOMAL_DOMINANT, diseaseId=OMIM:101600, hpoIds=[], seedGeneList=[], numberOfGenesToShow=0, outFileName=results/Pfeiffer-exomiser-results, outputFormat=[HTML], diseaseGeneFamilyName=, buildVersion=, buildTimestamp=}</pre>
                </div>

                <div class="tab-pane fade" id="filtering">
                    <table class="table">
                        <thead>
                            <tr>
                                <th>Filter</th>
                                <th>Report</th>
                                <th>Passed filter</th>
                                <th>Failed filter</th>
                            </tr>
                        </thead>
                        <tbody>
                            <tr th:each="report : ${filterReports}">
                                <td><b th:text="${report.filterType.shortName}">Frequency</b></td>
                                <td>
                                    <ul th:each="message : ${report.messages}" 
                                        th:if="${not #lists.isEmpty(report.messages)}" 
                                        th:text="${message}">Allele frequency &lt; 1.00 %</ul>
                                </td>
                                <td th:text="${report.passed}">16916</td>
                                <td th:text="${report.failed}">20793</td>
                            </tr>
                        </tbody>
                    </table>
                    <br/>
                </div>

                <div class="tab-pane fade" id="distribution">
                    <table class="table-striped table-condensed">
                        <thead>
                            <tr>
                                <th>Variant Type</th>    
                                <th:block th:each="sampleName: ${sampleNames}">
                                    <th th:text="${sampleName}">Slartibartfast</th>
                                </th:block>
                            </tr>
                        </thead>
                        <tbody>
                            <tr th:each="variantTypeCount: ${variantTypeCounters}">
                                <td th:text="${variantTypeCount.variantType}">MISSENSE</td>
                                <th:block th:each="count: ${variantTypeCount.sampleVariantTypeCounts}">
                                    <td th:text="${count}">1</td>
                                </th:block>
                            </tr>
                        </tbody>
                    </table>
                </div>

                <div class="tab-pane fade" id="errors" th:unless="${#lists.isEmpty(unAnalysedVarEvals)}">
                    <div class="row" th:each="variantEvaluation: ${unAnalysedVarEvals}"> 
                        <div class="col-sm-12">
                            <span class="label label-danger" th:text="${variantEvaluation.variantEffect}">MISSENSE</span>
                            <b th:text=" |${variantEvaluation.toGnomad} [${variantEvaluation.genotypeString}]|">chr10:g.123256215T>G [0/1]</b>
                            <b th:if="${variantEvaluation.isSymbolic()}"  th:text=" |${variantEvaluation.changeLengthString}|"> 10 Mb</b>
                            <b th:each="frequencyData: ${variantEvaluation.frequencyData}">
                                <a th:each="rsId: ${#objects.nullSafe(frequencyData.rsId, default)}"
                                   th:href="@{|https://www.ensembl.org/Homo_sapiens/snpview/?source=dbSNP;snp=${rsId}|}" th:text="${rsId}">rs141235720
                                </a>
                                <a th:each="rsId: ${#objects.nullSafe(frequencyData.rsId, default)}" 
                                   th:href="@{|https://www.ensembl.org/Homo_sapiens/Location/View/?v=${rsId};db=core;source=dbSNP;vdb=variation;|}"> (variation viewer)</a>
                            </b>
                        </div>
                    </div>
                </div>
                <!--/*-->

                <h2>Analyzed samples</h2>
                <!--                //if there is a multi-sample pedigree we want to see the analysis:
                                //HTMLTable.writePedigreeTable()-->
                <p>affected: red, parent of affected: light blue, unaffected: white</p>
                <table class="pedigree">
                    <tr>
                        <td id="r">Sample name: manuel</td>
                    </tr>
                    <tr>
                        <td>fam</td>
                        if (this.pedigree.isNthPersonAffected(i)) {
                        <td id="g">id</td>
                        } else if (this.pedigree.isNthPersonParentOfAffected(i)) {
                        <td id="b">id</td>
                        } else {
                        <td id="w">id</td>
                        }
                        <td>fathID</td>
                        <td>mothID</td>
                        <td>sex</td>
                        <td>disease</td>
                    </tr>
                </table>
                <!--*/-->

                <div class="tab-pane fade in active" id="prioritisation">

                    <div th:if="${geneResultsTruncated} == true" class="alert alert-info" role="alert">
                        For performance reasons only the top <span th:text="${@maxGenes}">300</span> out of a total of
                        <span th:text="${numCandidateGenes}">99999</span> candidate genes are shown here.
                        If you want to explore these other genes please try the offline version which you can
                        <a href="https://data.monarchinitiative.org/exomiser/"
                           class="alert-link">download from the FTP site</a>.
                    </div>

<!--                    <div class="panel panel-default" th:each="gene, iterStat: ${genes}">-->
<!--                        <div class="panel-heading">-->
<!--                            <div class="row">-->
<!--                                <div class="col-sm-3">-->
<!--                                    <h4 th:switch="${transcriptDb}"-->
<!--                                        th:with="geneIdentifier=${gene.getGeneIdentifier()}">-->
<!--                                        <b><a th:case="'UCSC'"-->
<!--                                              th:href="@{|http://www.ncbi.nlm.nih.gov/gene/${geneIdentifier.getEntrezId()}|}"-->
<!--                                              th:text="${geneIdentifier.getGeneSymbol()}">FGFR2</a></b>-->
<!--                                        <a th:case="'ENSEMBL'"-->
<!--                                           th:href="@{http://grch37.ensembl.org/Homo_sapiens/Gene/Summary(g=${geneIdentifier.getEnsemblId()})}"-->
<!--                                           th:text="${geneIdentifier.getGeneSymbol()}">FGFR2</a>-->
<!--                                        <a th:case="*"-->
<!--                                           th:href="@{|http://www.ncbi.nlm.nih.gov/gene/${geneIdentifier.getEntrezId()}|}"-->
<!--                                           th:text="${geneIdentifier.getGeneSymbol()}">FGFR2</a>-->
<!--                                    </h4>-->
<!--                                </div>-->
<!--                                <div class="col-sm-3">-->
<!--                                    <h4>Exomiser Score: <b th:text="${#numbers.formatDecimal(gene.combinedScore,1,3)}">1.005</b>-->
<!--                                        <data th:text="'(p=' + ${pValueFormatter.format(gene.pValue())} + ')'">(p=0.1e-4)</data>-->
<!--                                    </h4>-->
<!--                                </div>-->
<!--                                <div class="col-sm-3">-->
<!--                                    <h4>Phenotype Score: <b th:text="${#numbers.formatDecimal(gene.priorityScore,1,3)}">1.005</b>-->
<!--                                    </h4>-->
<!--                                </div>-->
<!--                                <div class="col-sm-3">-->
<!--                                    <h4>Variant Score: <b th:text="${#numbers.formatDecimal(gene.variantScore,1,3)}">1.005</b>-->
<!--                                    </h4>-->
<!--                                </div>-->
<!--                            </div>-->
<!--                        </div>-->
<!--                        <div class="panel-body">-->
<!--                            <div class="row">-->
<!--                                <div class="col-sm-12">-->
<!--                                    <b>Phenotype matches:</b>-->
<!--                                </div>-->
<!--                            </div>-->
<!--                            <div class="row">-->
<!--                                <div class="col-sm-12">-->
<!--                                    <th:block th:each="priorityResult: ${gene.priorityResults}"-->
<!--                                              th:utext="${priorityResult.value.HTMLCode}">-->
<!--                                        <a href="http://www.omim.org/entry/248450">Manitoba oculotrichoanal syndrome</a>-->
<!--                                    </th:block>-->
<!--                                </div>-->
<!--                            </div>-->
<!--                            <div class="row">-->
<!--                                <div class="col-sm-12">-->
<!--                                    <b>Top ranked variants:</b>-->
<!--                                </div>-->
<!--                            </div>-->
<!--                            <div class="row"-->
<!--                                 th:each="variantEvaluation: ${#lists.sort(gene.getPassedVariantEvaluations(), variantRankComparator)}">-->
<!--                                <div class="col-sm-12">-->
<!--                                    <span class="label label-danger"-->
<!--                                          th:text="${variantEvaluation.variantEffect}">MISSENSE</span>-->
<!--                                    <b th:text=" |${variantEvaluation.toGnomad} [${variantEvaluation.genotypeString}]|">-->
<!--                                        chr10:g.123256215T>G [0/1]</b>-->
<!--                                    <th:block th:each="rsId: ${variantEvaluation.frequencyData.rsId}"-->
<!--                                              th:if="${#strings.startsWith(rsId,'rs')}">-->
<!--                                        <a th:href="@{|https://www.ncbi.nlm.nih.gov/snp/${rsId}|}"-->
<!--                                           th:text="${rsId}">rs141235720-->
<!--                                        </a>-->
<!--                                    </th:block>-->
<!--                                </div>-->
<!--                                <div class="col-sm-12" th:if="${variantEvaluation.pathogenicityData.hasClinVarData()}">-->
<!--                                    <b>ClinVar: </b>-->
<!--                                    <th:block th:with="clinVarData = ${variantEvaluation.pathogenicityData.clinVarData()}"-->
<!--                                              th:switch="${clinVarData.getPrimaryInterpretation().toString()}">-->
<!--                                        <span th:case="'BENIGN'" class="label label-success"-->
<!--                                              th:text="${clinVarData.getPrimaryInterpretation()}">BENIGN</span>-->
<!--                                        <span th:case="'BENIGN_OR_LIKELY_BENIGN'" class="label label-success"-->
<!--                                              th:text="${clinVarData.getPrimaryInterpretation()}">BENIGN_OR_LIKELY_BENIGN</span>-->
<!--                                        <span th:case="'LIKELY_BENIGN'" class="label label-info"-->
<!--                                              th:text="${clinVarData.getPrimaryInterpretation()}">LIKELY_BENIGN</span>-->
<!--                                        <span th:case="'LIKELY_PATHOGENIC'" class="label label-warning"-->
<!--                                              th:text="${clinVarData.getPrimaryInterpretation()}">LIKELY_PATHOGENIC</span>-->
<!--                                        <span th:case="'PATHOGENIC_OR_LIKELY_PATHOGENIC'" class="label label-danger"-->
<!--                                              th:text="${clinVarData.getPrimaryInterpretation()}">PATHOGENIC_OR_LIKELY_PATHOGENIC</span>-->
<!--                                        <span th:case="'PATHOGENIC'" class="label label-danger"-->
<!--                                              th:text="${clinVarData.getPrimaryInterpretation()}">PATHOGENIC</span>-->
<!--                                        <span th:case="*" class="label label-default"-->
<!--                                              th:text="${clinVarData.getPrimaryInterpretation()}">UNCERTAIN_SIGNIFICANCE</span>-->
<!--                                        <a th:href="@{https://www.ncbi.nlm.nih.gov/clinvar/(term=${clinVarData.alleleId} + '[alleleid]')}"-->
<!--                                           th:text="'(' + ${clinVarData.getReviewStatus()} + ')'">-->
<!--                                            (criteria_provided,_single_submitter)-->
<!--                                        </a>-->
<!--                                    </th:block>-->
<!--                                </div>-->
<!--                                <div class="col-sm-6">-->
<!--                                    <b th:text="'Variant score: ' + ${#numbers.formatDecimal(variantEvaluation.getVariantScore(),1,3)}">Variant-->
<!--                                        score: 0.600</b>-->
<!--                                    <span class="label label-info" th:if="${variantEvaluation.contributesToGeneScore()}"-->
<!--                                          th:text="'CONTRIBUTING VARIANT'">CONTRIBUTING VARIANT</span>-->
<!--                                    <dl class="list-unstyled">-->
<!--                                        <dt>Transcripts:</dt>-->
<!--                                        <dd th:each="annotation: ${variantEvaluation.transcriptAnnotations}"-->
<!--                                            th:switch="${transcriptDb}">-->
<!--                                            <a th:case="'UCSC'"-->
<!--                                               th:href="@{http://genome.ucsc.edu/cgi-bin/hgTracks?db=hg19(position=${annotation.accession})}"-->
<!--                                               th:text="${gene.getGeneSymbol() + ':' + annotation.getAccession() + ':' + annotation.getHgvsCdna() + ':' + annotation.getHgvsProtein()}">FGFR2:uc001lfg.4:c.518A>C:p.E173A</a>-->
<!--                                            <a th:case="'ENSEMBL'"-->
<!--                                               th:href="@{http://grch37.ensembl.org/Homo_sapiens/Transcript/Summary(t=${annotation.accession})}"-->
<!--                                               th:text="${gene.getGeneSymbol() + ':' + annotation.getAccession() + ':' + annotation.getHgvsCdna() + ':' + annotation.getHgvsProtein()}">FGFR2:ENST00000429361.1:c.518A>C:p.E173A</a>-->
<!--                                            <a th:case="'REFSEQ'"-->
<!--                                               th:href="@{|https://www.ncbi.nlm.nih.gov/nuccore/${annotation.accession}|}"-->
<!--                                               th:text="${gene.getGeneSymbol() + ':' + annotation.getAccession() + ':' + annotation.getHgvsCdna() + ':' + annotation.getHgvsProtein()}">FGFR2:NM_000141:c.518A>C:p.E173A</a>-->
<!--                                        </dd>-->
<!--                                    </dl>-->
<!--                                </div>-->
<!--                                <div class="col-sm-3">-->
<!--                                    &lt;!&ndash;<b>Pathogenicity Score: 1.00</b>&ndash;&gt;-->
<!--                                    <dl class="list-unstyled"-->
<!--                                        th:each="pathogenicityData: ${variantEvaluation.pathogenicityData}">-->
<!--                                        <dt>Pathogenicity Data:</dt>-->
<!--                                        <dd th:each="mostPathogenicScore: ${pathogenicityData.mostPathogenicScore}"-->
<!--                                            th:text="${#strings.prepend(mostPathogenicScore.score, 'Best Score: ')}">-->
<!--                                            1.00-->
<!--                                        </dd>-->
<!--                                        <dd th:if="${#lists.isEmpty(pathogenicityData.pathogenicityScores)}">No-->
<!--                                            pathogenicity data-->
<!--                                        </dd>-->
<!--                                        <dd th:each="pathScore: ${pathogenicityData.pathogenicityScores}"-->
<!--                                            th:text="${pathScore}">Mutation Taster: 0.999 (P)-->
<!--                                        </dd>-->
<!--                                    </dl>-->
<!--                                </div>-->
<!--                                <div class="col-sm-3">-->
<!--                                    &lt;!&ndash;<b>Frequency Score: 0.95</b>&ndash;&gt;-->
<!--                                    <dl th:each="frequencyData: ${variantEvaluation.frequencyData}">-->
<!--                                        <dt>Frequency Data:</dt>-->
<!--                                        <dd th:if="${#lists.isEmpty(frequencyData.frequencies)}">No frequency-->
<!--                                            data-->
<!--                                        </dd>-->
<!--                                        <dd th:each="freq: ${frequencyData.frequencies}"-->
<!--                                            th:text="${freq.source.source} + ': ' + ${#numbers.formatDecimal(freq.frequency,1,4)} + '%'">-->
<!--                                            1000Genomes: 0.02%-->
<!--                                        </dd>-->
<!--                                    </dl>-->
<!--                                </div>-->
<!--                            </div>-->
<!--                        </div>-->
<!--                    </div>-->
<!--                </div>-->
                    <div class="panel panel-default" th:each="gene, iterStat: ${genes}">
                        <div th:id="${gene.getGeneIdentifier().getGeneSymbol()}" class="panel-heading">
                            <div class="row">
                                <div class="col-sm-3">
                                    <h4 th:switch="${transcriptDb}" th:with="geneIdentifier=${gene.getGeneIdentifier()}">
                                        <b><a th:case="'UCSC'"
                                              th:href="@{|https://www.ncbi.nlm.nih.gov/gene/${geneIdentifier.getEntrezId()}|}"
                                              th:text="${geneIdentifier.getGeneSymbol()}">FGFR2</a></b>
                                        <a th:case="'ENSEMBL'"
                                           th:href="@{|https://${ensemblAssembly}.ensembl.org/Homo_sapiens/Gene/Summary?g=${geneIdentifier.getEnsemblId()}|}"
                                           th:text="${geneIdentifier.getGeneSymbol()}">FGFR2</a>
                                        <a th:case="*"
                                           th:href="@{|https://www.ncbi.nlm.nih.gov/gene/${geneIdentifier.getEntrezId()}|}"
                                           th:text="${geneIdentifier.getGeneSymbol()}">FGFR2</a>
                                    </h4>
                                </div>
                                <div class="col-sm-3">
                                    <h4>Exomiser Score: <b th:text="${#numbers.formatDecimal(gene.getCombinedScore(), 1, 3)}">1.00</b>
                                        <data th:text="'(p=' + ${pValueFormatter.format(gene.pValue())} + ')'">(p=0.1e-4)</data>
                                    </h4>
                                </div>
                                <div class="col-sm-3">
                                    <h4>Phenotype Score: <b th:text="${#numbers.formatDecimal(gene.getPriorityScore(), 1, 3)}">1.00</b></h4>
                                </div>
                                <div class="col-sm-3">
                                    <h4>Variant Score: <b th:text="${#numbers.formatDecimal(gene.getVariantScore(), 1, 3)}">1.00</b></h4>
                                </div>
                            </div>
                        </div>
                        <div class="panel-body">
                            <div class="row">
                                <div class="col-sm-12">
                                    <b>Phenotype matches:</b>
                                </div>
                            </div>
                            <div class="row">
                                <div class="col-sm-12">
                                    <th:block th:each="priorityResult: ${gene.priorityResults}"
                                              th:utext="${priorityResult.value.HTMLCode}">
                                        <a href="https://www.omim.org/entry/248450">Manitoba oculotrichoanal syndrome</a>
                                    </th:block>
                                </div>
                            </div>

                            <div class="row"
                                 th:unless="${#sets.isEmpty(gene.getCompatibleInheritanceModes())}">
                                <div class="col-sm-12">
                                    <b>Gene scores under compatible inheritance modes:</b>
                                </div>
                            </div>

                            <div class="panel panel-default"
                                 th:each="geneScore: ${#lists.sort(gene.getGeneScores())}"
                                 th:if="${not #lists.isEmpty(geneScore.getContributingVariants())}">

                                <!--<div th:if="${not #lists.isEmpty(geneScore.getContributingVariants())}">-->
                                <div class="panel-heading">
                                    <div class="row">
                                        <div class="col-sm-3">
                                            <h4 th:text="${geneScore.getModeOfInheritance()}">
                                                <b><a th:case="'AUTOSOMAL_DOMINANT'"
                                                      th:text="'AD'">AUTOSOMAL_DOMINANT</a>
                                                </b>
                                                <!--<a th:case="'ENSEMBL'"-->
                                                <!--th:href="@{http://grch37.ensembl.org/Homo_sapiens/Gene/Summary(g=${geneIdentifier.getEnsemblId()})}"-->
                                                <!--th:text="${geneIdentifier.getGeneSymbol()}">FGFR2</a>-->
                                                <!--<a th:case="*"-->
                                                <!--th:href="@{|http://www.ncbi.nlm.nih.gov/gene/${geneIdentifier.getEntrezId()}|}"-->
                                                <!--th:text="${geneIdentifier.getGeneSymbol()}">FGFR2</a>-->
                                            </h4>
                                        </div>
                                        <div class="col-sm-3">
                                            <h4>Exomiser Score:
                                                <b th:text="${#numbers.formatDecimal(geneScore.getCombinedScore(), 1, 3)}">1.005</b>
                                                <data th:text="'(p=' + ${pValueFormatter.format(geneScore.pValue())} + ')'">(p=0.1e-4)</data>
                                            </h4>
                                        </div>
                                        <div class="col-sm-3">
                                            <h4>Phenotype Score: <b
                                                    th:text="${#numbers.formatDecimal(geneScore.getPhenotypeScore(), 1, 3)}">1.005</b>
                                            </h4>
                                        </div>
                                        <div class="col-sm-3">
                                            <h4>Variant Score: <b
                                                    th:text="${#numbers.formatDecimal(geneScore.getVariantScore(), 1, 3)}">1.005</b>
                                            </h4>
                                        </div>
                                    </div>
                                </div>
                                <div class="panel-body">
                                    <div class="row">
                                        <div class="col-sm-12">
                                            <b th:if="${#lists.isEmpty(geneScore.getCompatibleDiseaseMatches())}">
                                                No phenotype matches to diseases with this MOI.
                                            </b>
                                            <dl class="list-group"
                                                th:unless="${#lists.isEmpty(geneScore.getCompatibleDiseaseMatches())}">
                                                <!--                            th:each="diseaseModel: ${geneScore.getCompatibleDiseaseMatches()}">-->
                                                <dt>
                                                    Phenotype matches to diseases consistent with this MOI:
                                                </dt>
                                                <!--                            <dt th:each="diseaseModel: ${geneScore.getCompatibleDiseaseMatches()}"-->
                                                <!--                                th:text="Phenotypic similarity ${diseaseModel.getScore()} to ${diseaseModel.getModel().getDiseaseName()}">-->
                                                <!--                                Phenotypic similarity 0.95 to APERT SYNDROME-->
                                                <!--                            </dt>-->
                                                <th:block th:each="diseaseModel: ${geneScore.getCompatibleDiseaseMatches()}">
                                                    <dt th:text="|Phenotypic similarity ${#numbers.formatDecimal(diseaseModel.getScore(), 1, 3)} to ${diseaseModel.getModel().getId()} ${diseaseModel.getModel().getDiseaseName()}|">
                                                        Phenotypic similarity 0.957 to OMIM:101600 APERT SYNDROME
                                                    </dt>
                                                </th:block>
                                            </dl>
                                        </div>
                                    </div>
                                </div>
                                <div class="panel-body">
                                    <div class="row">
                                        <div class="col-sm-12">
                                            <b>Variants contributing to score:</b>
                                        </div>
                                    </div>
                                    <div class="row"
                                         th:each="acmgAssignment: ${geneScore.getAcmgAssignments()}"
                                         th:with="variantEvaluation = ${acmgAssignment.variantEvaluation()}">
                                        <div class="col-sm-12">
                                <span class="label label-danger"
                                      th:text="${variantEvaluation.variantEffect}">MISSENSE</span>
                                            <b th:text=" ${variantEvaluation.variantType}"> SNV</b>
                                            <b th:text=" |${variantEvaluation.toGnomad} [${variantEvaluation.genotypeString}]|">chr10:g.123256215T>G [0/1]</b>
                                            <b th:if="${variantEvaluation.isSymbolic()}"  th:text=" |${variantEvaluation.changeLengthString}|"> 10 Mb</b>
                                            <th:block th:each="varId: ${variantEvaluation.frequencyData.rsId}"
                                                      th:if="!${#strings.isEmpty(varId)}"
                                                      th:switch="${#strings.substring(varId, 0, 2)}">
                                                <a th:case="'rs'"
                                                   th:href="@{|https://www.ncbi.nlm.nih.gov/snp/${varId}|}"
                                                   th:text="${varId}">rs141235720
                                                </a>
                                                <a th:case="'es'"
                                                   th:href="@{|https://www.ncbi.nlm.nih.gov/dbvar/variants/${varId}|}"
                                                   th:text="${varId}">esv3097123
                                                </a>
                                                <a th:case="'ns'"
                                                   th:href="@{|https://www.ncbi.nlm.nih.gov/dbvar/variants/${varId}|}"
                                                   th:text="${varId}">nsv819538
                                                </a>
                                                <a th:case="'dg'"
                                                   th:href="@{|http://dgv.tcag.ca/dgv/app/variant(id=${varId})|}"
                                                   th:text="${varId}">dgv1891n106
                                                </a>
                                                <span th:case="*"
                                                      th:text="${varId}">12345
                                    </span>
                                            </th:block>
                                        </div>
                                        <div class="col-sm-12">
                                            <b>Exomiser ACMG: </b>
                                            <th:block th:with="acmgClassification = ${acmgAssignment.acmgClassification()}"
                                                      th:switch="${acmgClassification.toString()}">
                                    <span th:case="'BENIGN'" class="label label-success"
                                          th:text="${acmgClassification}">BENIGN</span>
                                                <span th:case="'LIKELY_BENIGN'" class="label label-info"
                                                      th:text="${acmgClassification}">LIKELY_BENIGN</span>
                                                <span th:case="'LIKELY_PATHOGENIC'" class="label label-warning"
                                                      th:text="${acmgClassification}">LIKELY_PATHOGENIC</span>
                                                <span th:case="'PATHOGENIC'" class="label label-danger"
                                                      th:text="${acmgClassification}">PATHOGENIC</span>
                                                <span th:case="'UNCERTAIN_SIGNIFICANCE'" class="label label-default"
                                                      th:text="${acmgClassification}">UNCERTAIN_SIGNIFICANCE</span>
                                                <span th:case="*" class="label label-default"
                                                      th:text="${acmgClassification}">NOT_AVAILABLE</span>
                                                <b
                                                        th:text="${acmgAssignment.acmgEvidence()}">
                                                    [PVS1, BP2]
                                                </b>
                                            </th:block>
                                        </div>
                                        <div class="col-sm-12" th:if="${variantEvaluation.pathogenicityData.hasClinVarData()}">
                                            <b>ClinVar: </b>
                                            <th:block th:with="clinVarData = ${variantEvaluation.pathogenicityData.clinVarData()}"
                                                      th:switch="${clinVarData.getPrimaryInterpretation().toString()}">
                                    <span th:case="'BENIGN'" class="label label-success"
                                          th:text="${clinVarData.getPrimaryInterpretation()}">BENIGN</span>
                                                <span th:case="'BENIGN_OR_LIKELY_BENIGN'" class="label label-success"
                                                      th:text="${clinVarData.getPrimaryInterpretation()}">BENIGN_OR_LIKELY_BENIGN</span>
                                                <span th:case="'LIKELY_BENIGN'" class="label label-info"
                                                      th:text="${clinVarData.getPrimaryInterpretation()}">LIKELY_BENIGN</span>
                                                <span th:case="'LIKELY_PATHOGENIC'" class="label label-warning"
                                                      th:text="${clinVarData.getPrimaryInterpretation()}">LIKELY_PATHOGENIC</span>
                                                <span th:case="'PATHOGENIC_OR_LIKELY_PATHOGENIC'" class="label label-danger"
                                                      th:text="${clinVarData.getPrimaryInterpretation()}">PATHOGENIC_OR_LIKELY_PATHOGENIC</span>
                                                <span th:case="'PATHOGENIC'" class="label label-danger"
                                                      th:text="${clinVarData.getPrimaryInterpretation()}">PATHOGENIC</span>
                                                <span th:case="*" class="label label-default"
                                                      th:text="${clinVarData.getPrimaryInterpretation()}">UNCERTAIN_SIGNIFICANCE</span>
                                                <a th:href="@{|https://www.ncbi.nlm.nih.gov/clinvar/variation/${clinVarData.variationId}|}"
                                                   th:text="'(' + ${clinVarData.getReviewStatus()} + ')'">
                                                    (criteria_provided,_single_submitter)
                                                </a>
                                                <span th:with="conflictingInterpretationCounts = ${clinVarData.conflictingInterpretationCounts}"
                                                      th:unless="${#maps.isEmpty(clinVarData.conflictingInterpretationCounts)}"
                                                      th:text="'(' + ${conflictingInterpretationsFormatter.format(conflictingInterpretationCounts)} + ')'">(P:3, LP:2, VUS:1)
                                                </span>
                                            </th:block>
                                        </div>
                                        <div class="col-sm-6">
                                            <b th:text="'Variant score: ' + ${#numbers.formatDecimal(variantEvaluation.getVariantScore(),1,3)}">Variant
                                                score: 0.600</b>
                                            <span class="label label-info"
                                                  th:if="${variantEvaluation.contributesToGeneScore()}"
                                                  th:text="'CONTRIBUTING VARIANT'">CONTRIBUTING VARIANT</span>
                                            <span class="label label-danger"
                                                  th:if="${variantEvaluation.isWhiteListed()}"
                                                  th:text="'WHITELIST VARIANT'">WHITELIST VARIANT</span>
                                            <dl class="list-unstyled">
                                                <dt>Transcripts:</dt>
                                                <dd th:each="annotation: ${variantEvaluation.transcriptAnnotations}"
                                                    th:switch="${transcriptDb}">
                                                    <a th:case="'UCSC'"
                                                       th:href="@{|https://genome.ucsc.edu/cgi-bin/hgTracks?db=${ucscAssembly}&position=${annotation.accession}|}"
                                                       th:text="${gene.getGeneSymbol() + ':' + annotation.getAccession() + ':' + annotation.getHgvsCdna() + ':' + annotation.getHgvsProtein()}">FGFR2:uc001lfg.4:c.518A>C:p.E173A</a>
                                                    <a th:case="'ENSEMBL'"
                                                       th:href="@{|https://${ensemblAssembly}.ensembl.org/Homo_sapiens/Transcript/Summary?t=${annotation.accession}|}"
                                                       th:text="${gene.getGeneSymbol() + ':' + annotation.getAccession() + ':' + annotation.getHgvsCdna() + ':' + annotation.getHgvsProtein()}">FGFR2:ENST00000429361.1:c.518A>C:p.E173A</a>
                                                    <a th:case="'REFSEQ'"
                                                       th:href="@{|https://www.ncbi.nlm.nih.gov/nuccore/${annotation.accession}|}"
                                                       th:text="${gene.getGeneSymbol() + ':' + annotation.getAccession() + ':' + annotation.getHgvsCdna() + ':' + annotation.getHgvsProtein()}">FGFR2:NM_000141:c.518A>C:p.E173A</a>
                                                </dd>
                                            </dl>
                                        </div>
                                        <div class="col-sm-3">
                                            <!--<b>Pathogenicity Score: 1.00</b>-->
                                            <dl class="list-unstyled"
                                                th:each="pathogenicityData: ${variantEvaluation.pathogenicityData}">
                                                <dt>Pathogenicity Data:</dt>
                                                <dd th:each="mostPathogenicScore: ${pathogenicityData.mostPathogenicScore}"
                                                    th:text="${#strings.prepend(mostPathogenicScore.score, 'Best Score: ')}">1.00
                                                </dd>
                                                <dd th:if="${#lists.isEmpty(pathogenicityData.pathogenicityScores)}">No
                                                    pathogenicity data
                                                </dd>
                                                <dd th:each="pathScore: ${pathogenicityData.pathogenicityScores}"
                                                    th:text="${pathScore}">Mutation Taster: 0.999 (P)
                                                </dd>
                                            </dl>
                                        </div>
                                        <div class="col-sm-3">
                                            <!--<b>Frequency Score: 0.95</b>-->
                                            <dl th:each="frequencyData: ${variantEvaluation.frequencyData}">
                                                <dt>Frequency Data:</dt>
                                                <dd th:if="${#lists.isEmpty(frequencyData.frequencies)}">No frequency data</dd>
                                                <dd th:each="freq: ${frequencyData.frequencies}"
                                                    th:text="${freq.source.source} + ': ' + ${#numbers.formatDecimal(freq.frequency,1,4)} + '%'">
                                                    1000Genomes: 0.02%
                                                </dd>
                                            </dl>
                                        </div>
                                    </div> <!-- end variant row -->
                                </div> <!-- end panel-body -->
                                <!--</div> &lt;!&ndash; end gene score section&ndash;&gt;-->
                            </div> <!-- end gene scores -->
                            <th:block th:if="${not #lists.isEmpty(gene.getNonContributingPassedVariantEvaluations)}">
                                <div class="row">
                                    <div class="col-sm-12">
                                        <b>Other passed variants:</b>
                                    </div>
                                </div>
                                <div class="row"
                                     th:each="variantEvaluation: ${#lists.sort(gene.getPassedVariantEvaluations(), variantRankComparator)}"
                                     th:if="not ${variantEvaluation.contributesToGeneScore()}">
                                    <div class="col-sm-12">
                                        <span class="label label-danger" th:text="${variantEvaluation.variantEffect}">MISSENSE</span>
                                        <b th:text=" ${variantEvaluation.variantType}"> SNV</b>
                                        <b th:text=" |${variantEvaluation.toGnomad} [${variantEvaluation.genotypeString}]|">chr10:g.123256215T>G [0/1]</b>
                                        <b th:if="${variantEvaluation.isSymbolic()}"  th:text=" |${variantEvaluation.changeLengthString}|"> 10 Mb</b>
                                        <th:block th:each="varId: ${variantEvaluation.frequencyData.rsId}"
                                                  th:if="!${#strings.isEmpty(varId)}"
                                                  th:switch="${#strings.substring(varId, 0, 2)}">
                                            <a th:case="'rs'"
                                               th:href="@{|https://www.ncbi.nlm.nih.gov/snp/${varId}|}"
                                               th:text="${varId}">rs141235720
                                            </a>
                                            <a th:case="'es'"
                                               th:href="@{|https://www.ncbi.nlm.nih.gov/dbvar/variants/${varId}|}"
                                               th:text="${varId}">esv3097123
                                            </a>
                                            <a th:case="'ns'"
                                               th:href="@{|https://www.ncbi.nlm.nih.gov/dbvar/variants/${varId}|}"
                                               th:text="${varId}">nsv819538
                                            </a>
                                            <a th:case="'dg'"
                                               th:href="@{|http://dgv.tcag.ca/dgv/app/variant(id=${varId})|}"
                                               th:text="${varId}">dgv1891n106
                                            </a>
                                            <span th:case="*"
                                                  th:text="${varId}">12345
                                    </span>
                                        </th:block>
                                    </div>
                                    <!--                    <div class="col-sm-12">-->
                                    <!--                        <b>Exomiser ACMG: </b>-->
                                    <!--                        <th:bloc  th:with="moi =${geneScore.getModeOfInheritance()} , acmgClassification = ${variantEvaluation.acmgAssignmentForMoi(${moi})}"-->
                                    <!--                                  th:switch="${acmgClassification.toString()}">-->
                                    <!--                                    <span th:case="'BENIGN'" class="label label-success"-->
                                    <!--                                          th:text="${acmgClassification}">BENIGN</span>-->
                                    <!--                            <span th:case="'LIKELY_BENIGN'" class="label label-info"-->
                                    <!--                                  th:text="${acmgClassification}">LIKELY_BENIGN</span>-->
                                    <!--                            <span th:case="'LIKELY_PATHOGENIC'" class="label label-warning"-->
                                    <!--                                  th:text="${acmgClassification}">LIKELY_PATHOGENIC</span>-->
                                    <!--                            <span th:case="'PATHOGENIC'" class="label label-danger"-->
                                    <!--                                  th:text="${acmgClassification}">PATHOGENIC</span>-->
                                    <!--                            <span th:case="'UNCERTAIN_SIGNIFICANCE'" class="label label-default"-->
                                    <!--                                  th:text="${acmgClassification}">UNCERTAIN_SIGNIFICANCE</span>-->
                                    <!--                            <span th:case="*" class="label label-default"-->
                                    <!--                                  th:text="${acmgClassification}">NOT_AVAILABLE</span>-->
                                    <!--                            <b-->
                                    <!--                                    th:text="${acmgClassification.acmgCriteria()}">-->
                                    <!--                                [PVS1, BP2]-->
                                    <!--                            </b>-->
                                    <!--                        </th:bloc>-->
                                    <!--                    </div>-->
                                    <div class="col-sm-12" th:if="${variantEvaluation.pathogenicityData.hasClinVarData()}">
                                        <b>ClinVar: </b>
                                        <th:block th:with="clinVarData = ${variantEvaluation.pathogenicityData.clinVarData()}"
                                                  th:switch="${clinVarData.getPrimaryInterpretation().toString()}">
                                            <span th:case="'BENIGN'" class="label label-success"
                                                  th:text="${clinVarData.getPrimaryInterpretation()}">BENIGN</span>
                                            <span th:case="'BENIGN_OR_LIKELY_BENIGN'" class="label label-success"
                                                  th:text="${clinVarData.getPrimaryInterpretation()}">BENIGN_OR_LIKELY_BENIGN</span>
                                            <span th:case="'LIKELY_BENIGN'" class="label label-info"
                                                  th:text="${clinVarData.getPrimaryInterpretation()}">LIKELY_BENIGN</span>
                                            <span th:case="'LIKELY_PATHOGENIC'" class="label label-warning"
                                                  th:text="${clinVarData.getPrimaryInterpretation()}">LIKELY_PATHOGENIC</span>
                                            <span th:case="'PATHOGENIC_OR_LIKELY_PATHOGENIC'" class="label label-danger"
                                                  th:text="${clinVarData.getPrimaryInterpretation()}">PATHOGENIC_OR_LIKELY_PATHOGENIC</span>
                                            <span th:case="'PATHOGENIC'" class="label label-danger"
                                                  th:text="${clinVarData.getPrimaryInterpretation()}">PATHOGENIC</span>
                                            <span th:case="*" class="label label-default"
                                                  th:text="${clinVarData.getPrimaryInterpretation()}">UNCERTAIN_SIGNIFICANCE</span>
                                            <a th:href="@{|https://www.ncbi.nlm.nih.gov/clinvar/variation/${clinVarData.variationId}|}"
                                               th:text="'(' + ${clinVarData.getReviewStatus()} + ')'">
                                                (criteria_provided,_single_submitter)
                                            </a>
                                            <span th:with="conflictingInterpretationCounts = ${clinVarData.conflictingInterpretationCounts}"
                                                  th:unless="${#maps.isEmpty(clinVarData.conflictingInterpretationCounts)}"
                                                  th:text="'(' + ${conflictingInterpretationsFormatter.format(conflictingInterpretationCounts)} + ')'">(P:3, LP:2, VUS:1)
                                            </span>
                                        </th:block>
                                    </div>
                                    <div class="col-sm-6">
                                        <b th:text="'Variant score: ' + ${#numbers.formatDecimal(variantEvaluation.getVariantScore(),1,3)}">Variant
                                            score: 0.600</b>
                                        <span class="label label-danger"
                                              th:if="${variantEvaluation.isWhiteListed()}"
                                              th:text="'WHITELIST VARIANT'">WHITELIST VARIANT</span>
                                        <dl class="list-unstyled">
                                            <dt>Transcripts:</dt>
                                            <dd th:each="annotation: ${variantEvaluation.transcriptAnnotations}"
                                                th:switch="${transcriptDb}">
                                                <a th:case="'UCSC'"
                                                   th:href="@{|https://genome.ucsc.edu/cgi-bin/hgTracks?db=${ucscAssembly}&position=${annotation.accession}|}"
                                                   th:text="${gene.getGeneSymbol() + ':' + annotation.getAccession() + ':' + annotation.getHgvsCdna() + ':' + annotation.getHgvsProtein()}">FGFR2:uc001lfg.4:c.518A>C:p.E173A</a>
                                                <a th:case="'ENSEMBL'"
                                                   th:href="@{|https://${ensemblAssembly}.ensembl.org/Homo_sapiens/Transcript/Summary?t=${annotation.accession}|}"
                                                   th:text="${gene.getGeneSymbol() + ':' + annotation.getAccession() + ':' + annotation.getHgvsCdna() + ':' + annotation.getHgvsProtein()}">FGFR2:ENST00000429361.1:c.518A>C:p.E173A</a>
                                                <a th:case="'REFSEQ'"
                                                   th:href="@{|https://www.ncbi.nlm.nih.gov/nuccore/${annotation.accession}|}"
                                                   th:text="${gene.getGeneSymbol() + ':' + annotation.getAccession() + ':' + annotation.getHgvsCdna() + ':' + annotation.getHgvsProtein()}">FGFR2:NM_000141:c.518A>C:p.E173A</a>
                                            </dd>
                                        </dl>
                                    </div>
                                    <div class="col-sm-3">
                                        <!--<b>Pathogenicity Score: 1.00</b>-->
                                        <dl class="list-unstyled"
                                            th:each="pathogenicityData: ${variantEvaluation.pathogenicityData}">
                                            <dt>Pathogenicity Data:</dt>
                                            <dd th:each="mostPathogenicScore: ${pathogenicityData.mostPathogenicScore}"
                                                th:text="${#strings.prepend(mostPathogenicScore.score, 'Best Score: ')}">1.00
                                            </dd>
                                            <dd th:if="${#lists.isEmpty(pathogenicityData.pathogenicityScores)}">No
                                                pathogenicity data
                                            </dd>
                                            <dd th:each="pathScore: ${pathogenicityData.pathogenicityScores}"
                                                th:text="${pathScore}">Mutation Taster: 0.999 (P)
                                            </dd>
                                        </dl>
                                    </div>
                                    <div class="col-sm-3">
                                        <!--<b>Frequency Score: 0.95</b>-->
                                        <dl th:each="frequencyData: ${variantEvaluation.frequencyData}">
                                            <dt>Frequency Data:</dt>
                                            <dd th:if="${#lists.isEmpty(frequencyData.frequencies)}">No frequency data</dd>
                                            <dd th:each="freq: ${frequencyData.frequencies}"
                                                th:text="${freq.source.source} + ': ' + ${#numbers.formatDecimal(freq.frequency,1,4)} + '%'">
                                                1000Genomes: 0.02%
                                            </dd>
                                        </dl>
                                    </div>
                                </div>
                            </th:block>
                        </div>
                    </div>

                <div class="tab-pane fade" id="about">
                    <p>Problems, suggestions, or comments? Please <a href="mailto:peter.robinson@jax.org">let us know</a></p>
                </div>
            </div>
        </div>

        <footer th:insert="~{footer :: footer}" class="navbar-default navbar-bottom"></footer>

        <script type="text/javascript" src="https://cdn.jsdelivr.net/webjars/jquery/2.1.4/jquery.min.js" th:src="@{/webjars/jquery/2.1.4/jquery.min.js}"></script>
        <script type="text/javascript" src="https://cdn.jsdelivr.net/webjars/bootstrap/3.3.7-1/js/bootstrap.min.js" th:href="@{/webjars/bootstrap/3.3.7-1/js/bootstrap.min.js}"></script>

        <script type ="text/javascript">
            $(document).ready(function () {
                $('#tabs a').click(function (e) {
                    $(this).tab('show');
                });
            });
        </script>
    </body>
</html>
