/*
 * The Exomiser - A tool to annotate and prioritize genomic variants
 *
 * Copyright (c) 2016-2021 Queen Mary University of London.
 * Copyright (c) 2012-2016 Charité Universitätsmedizin Berlin and Genome Research Ltd.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.monarchinitiative.exomiser.core.model;

import com.google.common.collect.ImmutableMap;
import de.charite.compbio.jannovar.annotation.VariantEffect;
import org.monarchinitiative.exomiser.core.model.frequency.Frequency;
import org.monarchinitiative.exomiser.core.model.frequency.FrequencyData;
import org.monarchinitiative.exomiser.core.model.frequency.FrequencySource;
import org.monarchinitiative.exomiser.core.model.pathogenicity.ClinVarData;
import org.monarchinitiative.exomiser.core.model.pathogenicity.PathogenicityData;
import org.monarchinitiative.exomiser.core.model.pathogenicity.PathogenicityScore;
import org.monarchinitiative.exomiser.core.model.pathogenicity.PathogenicitySource;
import org.monarchinitiative.exomiser.core.proto.AlleleProto;
import org.monarchinitiative.exomiser.core.proto.AlleleProto.AlleleKey;
import org.monarchinitiative.exomiser.core.proto.AlleleProto.AlleleProperties;
import org.monarchinitiative.exomiser.core.proto.AlleleProto.ClinVar;
import org.monarchinitiative.svart.GenomicVariant;

import java.util.*;

import static org.monarchinitiative.exomiser.core.model.frequency.FrequencySource.*;
import static org.monarchinitiative.exomiser.core.model.pathogenicity.PathogenicitySource.*;

/**
 * @author Jules Jacobsen <j.jacobsen@qmul.ac.uk>
 * @since 10.1.0
 */
public class AlleleProtoAdaptor {

    // These maps are constant look-ups for keys in the AlleleProto.AlleleProperties propertiesMap which was generated by the
    // genome-data module. The keys are AlleleProperty string values.
    private static final Map<String, FrequencySource> FREQUENCY_SOURCE_MAP = new ImmutableMap.Builder<String, FrequencySource>()
            .put("KG", THOUSAND_GENOMES)
            .put("TOPMED", TOPMED)
            .put("UK10K", UK10K)

            .put("ESP_AA", ESP_AFRICAN_AMERICAN)
            .put("ESP_EA", ESP_EUROPEAN_AMERICAN)
            .put("ESP_ALL", ESP_ALL)

            .put("EXAC_AFR", EXAC_AFRICAN_INC_AFRICAN_AMERICAN)
            .put("EXAC_AMR", EXAC_AMERICAN)
            .put("EXAC_EAS", EXAC_EAST_ASIAN)
            .put("EXAC_FIN", EXAC_FINNISH)
            .put("EXAC_NFE", EXAC_NON_FINNISH_EUROPEAN)
            .put("EXAC_OTH", EXAC_OTHER)
            .put("EXAC_SAS", EXAC_SOUTH_ASIAN)

            .put("GNOMAD_E_AFR", GNOMAD_E_AFR)
            .put("GNOMAD_E_AMR", GNOMAD_E_AMR)
            .put("GNOMAD_E_ASJ", GNOMAD_E_ASJ)
            .put("GNOMAD_E_EAS", GNOMAD_E_EAS)
            .put("GNOMAD_E_FIN", GNOMAD_E_FIN)
            .put("GNOMAD_E_NFE", GNOMAD_E_NFE)
            .put("GNOMAD_E_OTH", GNOMAD_E_OTH)
            .put("GNOMAD_E_SAS", GNOMAD_E_SAS)

            .put("GNOMAD_G_AFR", GNOMAD_G_AFR)
            .put("GNOMAD_G_AMR", GNOMAD_G_AMR)
            .put("GNOMAD_G_ASJ", GNOMAD_G_ASJ)
            .put("GNOMAD_G_EAS", GNOMAD_G_EAS)
            .put("GNOMAD_G_FIN", GNOMAD_G_FIN)
            .put("GNOMAD_G_NFE", GNOMAD_G_NFE)
            .put("GNOMAD_G_OTH", GNOMAD_G_OTH)
            .put("GNOMAD_G_SAS", GNOMAD_G_SAS)
            .build();

    private static final Map<String, PathogenicitySource> PATHOGENICITY_SOURCE_MAP = new ImmutableMap.Builder<String, PathogenicitySource>()
            .put("POLYPHEN", POLYPHEN)
            .put("MUT_TASTER", MUTATION_TASTER)
            .put("SIFT", SIFT)
            .put("CADD", CADD)
            .put("REMM", REMM)
            .put("REVEL", REVEL)
            .put("MCAP", M_CAP)
            .put("MPC", MPC)
            .put("MVP", MVP)
            .put("PRIMATE_AI", PRIMATE_AI)
            .build();

    private AlleleProtoAdaptor() {
        //un-instantiable utility class
    }

    // This would make sense to have this here rather than having similar functionality in the MvStoreUtil
    // and the VariantKeyGenerator
    public static AlleleKey toAlleleKey(GenomicVariant variant) {
        return AlleleKey.newBuilder()
                .setChr(variant.contigId())
                .setPosition(variant.start())
                .setRef(variant.ref())
                .setAlt(variant.alt())
                .build();
    }

    public static FrequencyData toFrequencyData(AlleleProperties alleleProperties) {
        if (alleleProperties.equals(AlleleProperties.getDefaultInstance())) {
            return FrequencyData.empty();
        }
        List<Frequency> frequencies = parseFrequencyData(alleleProperties.getPropertiesMap());
        return FrequencyData.of(alleleProperties.getRsId(), frequencies);
    }

    private static List<Frequency> parseFrequencyData(Map<String, Float> values) {
        List<Frequency> frequencies = new ArrayList<>(values.size());
        for (Map.Entry<String, Float> field : values.entrySet()) {
            String key = field.getKey();
            if (FREQUENCY_SOURCE_MAP.containsKey(key)) {
                FrequencySource source = FREQUENCY_SOURCE_MAP.get(key);
                float value = field.getValue();
                frequencies.add(Frequency.of(source, value));
            }
        }
        return frequencies;
    }

    public static PathogenicityData toPathogenicityData(AlleleProperties alleleProperties) {
        if (alleleProperties.equals(AlleleProperties.getDefaultInstance())) {
            return PathogenicityData.empty();
        }
        List<PathogenicityScore> pathogenicityScores = parsePathogenicityData(alleleProperties.getPropertiesMap());
        ClinVarData clinVarData = toClinVarData(alleleProperties.getClinVar());
        return PathogenicityData.of(clinVarData, pathogenicityScores);
    }

    private static List<PathogenicityScore> parsePathogenicityData(Map<String, Float> values) {
        List<PathogenicityScore> pathogenicityScores = new ArrayList<>();
        for (Map.Entry<String, Float> field : values.entrySet()) {
            String key = field.getKey();
            if(PATHOGENICITY_SOURCE_MAP.containsKey(key)) {
                PathogenicitySource source = PATHOGENICITY_SOURCE_MAP.get(key);
                float score = field.getValue();
                pathogenicityScores.add(PathogenicityScore.of(source, score));
            }
        }
        return pathogenicityScores;
    }

    public static ClinVarData toClinVarData(ClinVar clinVar) {
        if (clinVar.equals(clinVar.getDefaultInstanceForType())) {
            return ClinVarData.empty();
        }
        ClinVarData.Builder builder = ClinVarData.builder();
        builder.alleleId(clinVar.getAlleleId());
        builder.variationId(clinVar.getVariationId());
        builder.primaryInterpretation(toClinSig(clinVar.getPrimaryInterpretation()));
        builder.conflictingInterpretationCounts(toConflictCounts(clinVar.getClinSigCountsMap()));
        builder.secondaryInterpretations(toClinSigSet(clinVar.getSecondaryInterpretationsList()));
        builder.includedAlleles(getToIncludedAlleles(clinVar.getIncludedAllelesMap()));
        builder.reviewStatus(clinVar.getReviewStatus());
        builder.geneSymbol(clinVar.getGeneSymbol());
        builder.variantEffect(toVariantEffect(clinVar.getVariantEffect()));
        builder.hgvsCdna(clinVar.getHgvsCdna());
        builder.hgvsProtein(clinVar.getHgvsProtein());
        return builder.build();
    }

    private static Map<ClinVarData.ClinSig, Integer> toConflictCounts(Map<String, Integer> clinSigCountsMap) {
        if (clinSigCountsMap.isEmpty()) {
            return Collections.emptyMap();
        }
        Map<ClinVarData.ClinSig, Integer> converted = new EnumMap<>(ClinVarData.ClinSig.class);
        for (Map.Entry<String, Integer> included : clinSigCountsMap.entrySet()) {
            converted.put(ClinVarData.ClinSig.valueOf(included.getKey()), included.getValue());
        }
        return converted;
    }

    private static Map<String, ClinVarData.ClinSig> getToIncludedAlleles(Map<String, ClinVar.ClinSig> includedAllelesMap) {
        if (includedAllelesMap.isEmpty()) {
            return Collections.emptyMap();
        }
        Map<String, ClinVarData.ClinSig> converted = new HashMap<>(includedAllelesMap.size());

        for (Map.Entry<String, ClinVar.ClinSig> included : includedAllelesMap.entrySet()) {
            converted.put(included.getKey(), toClinSig(included.getValue()));
        }

        return converted;
    }

    private static Set<ClinVarData.ClinSig> toClinSigSet(List<ClinVar.ClinSig> protoClinSigs) {
        if (protoClinSigs.isEmpty()) {
            return Collections.emptySet();
        }
        Set<ClinVarData.ClinSig> converted = new HashSet<>(protoClinSigs.size());
        for (ClinVar.ClinSig protoClinSig : protoClinSigs) {
            converted.add(toClinSig(protoClinSig));
        }
        return converted;
    }

    private static ClinVarData.ClinSig toClinSig(ClinVar.ClinSig protoClinSig) {
        return switch (protoClinSig) {
            case BENIGN -> ClinVarData.ClinSig.BENIGN;
            case BENIGN_OR_LIKELY_BENIGN -> ClinVarData.ClinSig.BENIGN_OR_LIKELY_BENIGN;
            case LIKELY_BENIGN -> ClinVarData.ClinSig.LIKELY_BENIGN;
            case UNCERTAIN_SIGNIFICANCE -> ClinVarData.ClinSig.UNCERTAIN_SIGNIFICANCE;
            case LIKELY_PATHOGENIC -> ClinVarData.ClinSig.LIKELY_PATHOGENIC;
            case PATHOGENIC_OR_LIKELY_PATHOGENIC -> ClinVarData.ClinSig.PATHOGENIC_OR_LIKELY_PATHOGENIC;
            case PATHOGENIC -> ClinVarData.ClinSig.PATHOGENIC;
            case CONFLICTING_PATHOGENICITY_INTERPRETATIONS ->
                    ClinVarData.ClinSig.CONFLICTING_PATHOGENICITY_INTERPRETATIONS;
            case AFFECTS -> ClinVarData.ClinSig.AFFECTS;
            case ASSOCIATION -> ClinVarData.ClinSig.ASSOCIATION;
            case DRUG_RESPONSE -> ClinVarData.ClinSig.DRUG_RESPONSE;
            case OTHER -> ClinVarData.ClinSig.OTHER;
            case PROTECTIVE -> ClinVarData.ClinSig.PROTECTIVE;
            case RISK_FACTOR -> ClinVarData.ClinSig.RISK_FACTOR;
            default -> ClinVarData.ClinSig.NOT_PROVIDED;
        };
    }

    private static VariantEffect toVariantEffect(AlleleProto.VariantEffect clinVarVariantEffect) {
        return switch (clinVarVariantEffect) {
            case SEQUENCE_VARIANT -> VariantEffect.SEQUENCE_VARIANT;
            case CHROMOSOME_NUMBER_VARIATION -> VariantEffect.CHROMOSOME_NUMBER_VARIATION;
            case TRANSCRIPT_ABLATION -> VariantEffect.TRANSCRIPT_ABLATION;
            case EXON_LOSS_VARIANT -> VariantEffect.EXON_LOSS_VARIANT;
            case INVERSION -> VariantEffect.INVERSION;
            case INSERTION -> VariantEffect.INSERTION;
            case TRANSLOCATION -> VariantEffect.TRANSLOCATION;
            case FRAMESHIFT_ELONGATION -> VariantEffect.FRAMESHIFT_ELONGATION;
            case FRAMESHIFT_TRUNCATION -> VariantEffect.FRAMESHIFT_TRUNCATION;
            case FRAMESHIFT_VARIANT -> VariantEffect.FRAMESHIFT_VARIANT;
            case INTERNAL_FEATURE_ELONGATION -> VariantEffect.INTERNAL_FEATURE_ELONGATION;
            case FEATURE_TRUNCATION -> VariantEffect.FEATURE_TRUNCATION;
            case TRANSCRIPT_AMPLIFICATION -> VariantEffect.TRANSCRIPT_AMPLIFICATION;
            case COPY_NUMBER_CHANGE -> VariantEffect.COPY_NUMBER_CHANGE;
            case MNV -> VariantEffect.MNV;
            case COMPLEX_SUBSTITUTION -> VariantEffect.COMPLEX_SUBSTITUTION;
            case STOP_GAINED -> VariantEffect.STOP_GAINED;
            case STOP_LOST -> VariantEffect.STOP_LOST;
            case START_LOST -> VariantEffect.START_LOST;
            case SPLICE_ACCEPTOR_VARIANT -> VariantEffect.SPLICE_ACCEPTOR_VARIANT;
            case SPLICE_DONOR_VARIANT -> VariantEffect.SPLICE_DONOR_VARIANT;
            case RARE_AMINO_ACID_VARIANT -> VariantEffect.RARE_AMINO_ACID_VARIANT;
            case MISSENSE_VARIANT -> VariantEffect.MISSENSE_VARIANT;
            case INFRAME_INSERTION -> VariantEffect.INFRAME_INSERTION;
            case DISRUPTIVE_INFRAME_INSERTION -> VariantEffect.DISRUPTIVE_INFRAME_INSERTION;
            case INFRAME_DELETION -> VariantEffect.INFRAME_DELETION;
            case DISRUPTIVE_INFRAME_DELETION -> VariantEffect.DISRUPTIVE_INFRAME_DELETION;
            case FIVE_PRIME_UTR_TRUNCATION -> VariantEffect.FIVE_PRIME_UTR_TRUNCATION;
            case THREE_PRIME_UTR_TRUNCATION -> VariantEffect.THREE_PRIME_UTR_TRUNCATION;
            case SPLICE_REGION_VARIANT -> VariantEffect.SPLICE_REGION_VARIANT;
            case STOP_RETAINED_VARIANT -> VariantEffect.STOP_RETAINED_VARIANT;
            case INITIATOR_CODON_VARIANT -> VariantEffect.INITIATOR_CODON_VARIANT;
            case SYNONYMOUS_VARIANT -> VariantEffect.SYNONYMOUS_VARIANT;
            case CODING_TRANSCRIPT_INTRON_VARIANT -> VariantEffect.CODING_TRANSCRIPT_INTRON_VARIANT;
            case FIVE_PRIME_UTR_PREMATURE_START_CODON_GAIN_VARIANT ->
                    VariantEffect.FIVE_PRIME_UTR_PREMATURE_START_CODON_GAIN_VARIANT;
            case FIVE_PRIME_UTR_EXON_VARIANT -> VariantEffect.FIVE_PRIME_UTR_EXON_VARIANT;
            case THREE_PRIME_UTR_EXON_VARIANT -> VariantEffect.THREE_PRIME_UTR_EXON_VARIANT;
            case FIVE_PRIME_UTR_INTRON_VARIANT -> VariantEffect.FIVE_PRIME_UTR_INTRON_VARIANT;
            case THREE_PRIME_UTR_INTRON_VARIANT -> VariantEffect.THREE_PRIME_UTR_INTRON_VARIANT;
            case NON_CODING_TRANSCRIPT_EXON_VARIANT -> VariantEffect.NON_CODING_TRANSCRIPT_EXON_VARIANT;
            case NON_CODING_TRANSCRIPT_INTRON_VARIANT -> VariantEffect.NON_CODING_TRANSCRIPT_INTRON_VARIANT;
            case DIRECT_TANDEM_DUPLICATION -> VariantEffect.DIRECT_TANDEM_DUPLICATION;
            case MOBILE_ELEMENT_DELETION -> VariantEffect.MOBILE_ELEMENT_DELETION;
            case MOBILE_ELEMENT_INSERTION -> VariantEffect.MOBILE_ELEMENT_INSERTION;
            case UPSTREAM_GENE_VARIANT -> VariantEffect.UPSTREAM_GENE_VARIANT;
            case DOWNSTREAM_GENE_VARIANT -> VariantEffect.DOWNSTREAM_GENE_VARIANT;
            case INTERGENIC_VARIANT -> VariantEffect.INTERGENIC_VARIANT;
            case TFBS_ABLATION -> VariantEffect.TFBS_ABLATION;
            case TFBS_AMPLIFICATION -> VariantEffect.TFBS_AMPLIFICATION;
            case TF_BINDING_SITE_VARIANT -> VariantEffect.TF_BINDING_SITE_VARIANT;
            case REGULATORY_REGION_VARIANT -> VariantEffect.REGULATORY_REGION_VARIANT;
            case REGULATORY_REGION_ABLATION -> VariantEffect.REGULATORY_REGION_ABLATION;
            case REGULATORY_REGION_AMPLIFICATION -> VariantEffect.REGULATORY_REGION_AMPLIFICATION;
            case CONSERVED_INTRON_VARIANT -> VariantEffect.CONSERVED_INTRON_VARIANT;
            case INTRAGENIC_VARIANT -> VariantEffect.INTRAGENIC_VARIANT;
            case CONSERVED_INTERGENIC_VARIANT -> VariantEffect.CONSERVED_INTERGENIC_VARIANT;
            case STRUCTURAL_VARIANT -> VariantEffect.STRUCTURAL_VARIANT;
            case CODING_SEQUENCE_VARIANT -> VariantEffect.CODING_SEQUENCE_VARIANT;
            case INTRON_VARIANT -> VariantEffect.INTRON_VARIANT;
            case EXON_VARIANT -> VariantEffect.EXON_VARIANT;
            case SPLICING_VARIANT -> VariantEffect.SPLICING_VARIANT;
            case MIRNA -> VariantEffect.MIRNA;
            case CODING_TRANSCRIPT_VARIANT -> VariantEffect.CODING_TRANSCRIPT_VARIANT;
            case NON_CODING_TRANSCRIPT_VARIANT -> VariantEffect.NON_CODING_TRANSCRIPT_VARIANT;
            case UNRECOGNIZED -> VariantEffect.SEQUENCE_VARIANT;
        };
    }
}
