============
Publications
============

| **Evaluation of phenotype-driven gene prioritization methods for Mendelian diseases.**
| Jacobsen JOB, Kelly C, Cipriani V, Robinson PN, Smedley D.
| *Brief Bioinformatics. 2022 Sept 20; 23(5) bbac188*
| PMID: 35595299 DOI:`10.1093/bib/bbac188 <https://www.doi.org/10.1093/bib/bbac188>`_


| **Phenotype-aware prioritisation of rare Mendelian disease variants.**
| Kelly C, Szabo A, Pontikos N, Arno G, Robinson PN, Jacobsen JOB, Smedley D, Cipriani V.
| *Trends Genet. 2022 Aug 4; 38(12), 1271-1283*
| PMID:35934592 DOI:`10.1016/j.tig.2022.07.002 <https:www.doi.org/10.1016/j.tig.2022.07.002>`_


| **Phenotype-driven approaches to enhance variant prioritization and diagnosis of rare disease.**
| Jacobsen JOB, Kelly C, Cipriani V, Genomics England Research Consortium, Mungall CJ, Reese J, Danis D, Robinson PN, Smedley D.
| *Hum Mutat. 2022 April 7; 43(8), 1071-1081*
| PMID:35391505 DOI:`10.1002/humu.24380 <https://doi.org/10.1002/humu.24380>`_


| **100,000 Genomes Pilot on Rare-Disease Diagnosis in Health Care - Preliminary Report**
| 100,000 Genomes Project Pilot Investigators; Smedley D, ... Caulfield M.
| *N Engl J Med. 2021 Nov 11;385(20):1868-1880*
| PMID:34758253 DOI:`10.1056/NEJMoa2035790 <https://doi.org/10.1056/NEJMoa2035790>`_


| **An Improved Phenotype-Driven Tool for Rare Mendelian Variant Prioritization: Benchmarking Exomiser on Real Patient Whole-Exome Data.**
| Cipriani V, Pontikos N, Arno G, Sergouniotis PI, Lenassi E, Thawong P, Danis D, Michaelides M, Webster AR, Moore AT, Robinson PN, Jacobsen JOB, Smedley D.
| *Genes (Basel). 2020 Apr 23;11(4):460*
| PMID:32340307 DOI:`10.3390/genes11040460 <https://doi.org/10.3390/genes11040460>`_


| **A Whole-Genome Analysis Framework for Effective Identification of Pathogenic Regulatory Variants in Mendelian Disease.**
| Smedley D, Schubach M, Jacobsen JOB, Köhler S, Zemojtel T, Spielmann M, Jäger M, Hochheiser H, Washington NL, McMurry JA, Haendel MA, Mungall CJ, Lewis SE, Groza T, Valentini G, Robinson PN.
| *The American Journal of Human Genetics 2016;99;3;595-606*
| PMID:27569544 DOI:`10.1016/j.ajhg.2016.07.005 <https://doi.org/10.1016/j.ajhg.2016.07.005>`_


| **Phenotype-driven strategies for exome prioritization of human Mendelian disease genes.**
| Smedley D, Robinson PN.
| *Genome Medicine 2015, 7(1):81*
| PMID:26229552 DOI:`10.1186/s13073-015-0199-2 <https://doi.org/10.1186/s13073-015-0199-2>`_


| **Next-generation diagnostics and disease-gene discovery with the Exomiser.**
| Smedley D, Jacobsen JO, Jäger M, Köhler S, Holtgrewe M et al.
| *Nature protocols 2015;10;12;2004-15*
| PMID:26562621 DOI:`10.1038/nprot.2015.124 <https://doi.org/10.1038/nprot.2015.124>`_


| **Computational evaluation of exome sequence data using human and model organism phenotypes improves diagnostic efficiency.**
| Bone WP, Washington NL, Buske OJ, Adams DR, Davis J et al.
| *Genetics in medicine : official journal of the American College of Medical Genetics 2015*
| PMID:26562225 DOI:`10.1038/gim.2015.137 <https://doi.org/10.1038/gim.2015.137>`_


| **Walking the interactome for candidate prioritization in exome sequencing studies of Mendelian diseases.**
| Smedley D, Köhler S, Czeschik JC, Amberger J, Bocchini C et al.
| *Bioinformatics (Oxford, England) 2014;30;22;3215-22*
| PMID:25078397 DOI:`10.1093/bioinformatics/btu508 <https://doi.org/10.1093/bioinformatics/btu508>`_


| **Effective diagnosis of genetic disease by computational phenotype analysis of the disease-associated genome.**
| Zemojtel T, Köhler S, Mackenroth L, Jäger M, Hecht J et al.
| *Science translational medicine 2014;6;252;252ra123*
| PMID:25186178 DOI:`10.1126/scitranslmed.3009262 <https://doi.org/10.1126/scitranslmed.3009262>`_


| **Improved exome prioritization of disease genes through cross-species phenotype comparison.**
| Robinson PN, Köhler S, Oellrich A, Sanger Mouse Genetics Project, Wang K et al.
| *Genome research 2014;24;2;340-8*
| PMID:24162188 DOI:`10.1101/gr.160325.113 <https://doi.org/10.1101/gr.160325.113>`_


Collaborations
==============

Publications for other projects and resources in which the Exomiser is used as a component:

| **The RD-Connect Genome-Phenome Analysis Platform: Accelerating diagnosis, research, and gene discovery for rare diseases**
| Steven Laurie, Davide Piscia, Leslie Matalonga, Alberto Corvó, Marcos Fernández-Callejo, Carles Garcia-Linares, Carles Hernandez-Ferrer, Cristina Luengo, Inés Martínez … See all authors
| *Human Mutation 2022,43(6):717-733*
| PMID: 35178824 DOI:`10.1002/humu.24353 <https://doi.org/10.1002/humu.24353>`_


| **Defining Disease, Diagnosis, and Translational Medicine within a Homeostatic Perturbation Paradigm: The National Institutes of Health Undiagnosed Diseases Program Experience.**
| Gall T, Valkanas E, BelloC, Markello T, Adams C, Bone WP, Brandt AJ, Brazill JM, Carmichael L, Davids M, Davis J, Diaz-Perez Z, Draper D, Elson J, Flynn ED, Godfrey R, Groden C, Hsieh CK, Fischer R, Golas GA, Guzman J, Huang Y, Kane MS, Lee E, Li C, Links AE, Maduro V, Malicdan MCV, Malik FS, Nehrebecky M, Park J, Pemberton P, Schaffer K, Simeonov D, Sincan M, Smedley D, Valivullah Z, Wahl C, Washington N, Wolfe LA, Xu K, Zhu Y, Gahl WA, Tifft CJ, Toro C, Adams DR, He M, Robinson PN, Haendel MA, Zhai RG, Boerkoel CF.
| *Front Med (Lausanne). 2017;4 62*
| PMID: 28603714 DOI:`10.3389/fmed.2017.00062 <http://doi.org/10.3389/fmed.2017.00062>`_


| **PhenomeCentral: A Portal for Phenotypic and Genotypic Matchmaking of Patients with Rare Genetic Diseases.**
| Orion J. Buske, Marta Girdea, Sergiu Dumitriu, Bailey Gallinger, Taila Hartley, Heather Trang, Andriy Misyura, Tal Friedman, Chandree Beaulieu, William P. Bone, Amanda E. Links, Nicole L. Washington, Melissa A. Haendel, Peter N. Robinson, Cornelius F. Boerkoel, David Adams, William A. Gahl, Kym M. Boycott, Michael Brudno.
| *Human Mutation 2015, 36(10):931-940*
| PMID: 26251998 DOI:`10.1002/humu.22851 <http://doi.org/10.1002/humu.22851>`_


| **Phenopolis: an open platform for harmonization and analysis of genetic and phenotypic data.**
| Nikolas Pontikos, Jing Yu, Ismail Moghul, Lucy Withington, Fiona Blanco-Kelly, Tom Vulliamy, Tsz Lun, Ernest Wong, Cian Murphy, Valentina Cipriani, Alessia Fiorentino, Gavin Arno, Daniel Greene, Julius OB Jacobsen, Tristan Clark, David S. Gregory, Andrea M. Nemeth, Stephanie Halford, Chris F. Inglehearn, Susan Downes, Graeme C. Black, Andrew R. Webster, Alison J. Hardcastle, UKIRDC, Vincent Plagnol.
| *Bioinformatics 2017,33(15):2421-2423*
| PMID: 28334266 DOI:`10.1093/bioinformatics/btx147 <http://doi.org/10.1093/bioinformatics/btx147>`_

