============
Publications
============

A Whole-Genome Analysis Framework for Effective Identification of Pathogenic Regulatory Variants in Mendelian Disease. Smedley D, Schubach M, Jacobsen JOB, Köhler S, Zemojtel T, Spielmann M, Jäger M, Hochheiser H, Washington NL, McMurry JA, Haendel MA, Mungall CJ, Lewis SE, Groza T, Valentini G, Robinson PN.
The American Journal of Human Genetics 2016;99;3;595-606
PMID:27569544 DOI:10.1016/j.ajhg.2016.07.005

Phenotype-driven strategies for exome prioritization of human Mendelian disease genes. Smedley D, Robinson PN.
Genome Medicine 2015, 7(1):81
PMID:26229552 DOI:10.1186/s13073-015-0199-2

Next-generation diagnostics and disease-gene discovery with the Exomiser. Smedley D, Jacobsen JO, Jäger M, Köhler S, Holtgrewe M et al.
Nature protocols 2015;10;12;2004-15
PMID:26562621 DOI:10.1038/nprot.2015.124

Computational evaluation of exome sequence data using human and model organism phenotypes improves diagnostic efficiency. Bone WP, Washington NL, Buske OJ, Adams DR, Davis J et al.
Genetics in medicine : official journal of the American College of Medical Genetics 2015
PMID:26562225 DOI:10.1038/gim.2015.137

Walking the interactome for candidate prioritization in exome sequencing studies of Mendelian diseases. Smedley D, Köhler S, Czeschik JC, Amberger J, Bocchini C et al.
Bioinformatics (Oxford, England) 2014;30;22;3215-22
PMID:25078397 DOI:10.1093/bioinformatics/btu508

Effective diagnosis of genetic disease by computational phenotype analysis of the disease-associated genome. Zemojtel T, Köhler S, Mackenroth L, Jäger M, Hecht J et al.
Science translational medicine 2014;6;252;252ra123
PMID:25186178 DOI:10.1126/scitranslmed.3009262

Improved exome prioritization of disease genes through cross-species phenotype comparison. Robinson PN, Köhler S, Oellrich A, Sanger Mouse Genetics Project, Wang K et al.
Genome research 2014;24;2;340-8
PMID:24162188 DOI:10.1101/gr.160325.113

Publications for other projects and resources in which the Exomiser is used as a component:

Defining Disease, Diagnosis, and Translational Medicine within a Homeostatic Perturbation Paradigm: The National Institutes of Health Undiagnosed Diseases Program Experience. Gall T, Valkanas E, BelloC, Markello T, Adams C, Bone WP, Brandt AJ, Brazill JM, Carmichael L, Davids M, Davis J, Diaz-Perez Z, Draper D, Elson J, Flynn ED, Godfrey R, Groden C, Hsieh CK, Fischer R, Golas GA, Guzman J, Huang Y, Kane MS, Lee E, Li C, Links AE, Maduro V, Malicdan MCV, Malik FS, Nehrebecky M, Park J, Pemberton P, Schaffer K, Simeonov D, Sincan M, Smedley D, Valivullah Z, Wahl C, Washington N, Wolfe LA, Xu K, Zhu Y, Gahl WA, Tifft CJ, Toro C, Adams DR, He M, Robinson PN, Haendel MA, Zhai RG, Boerkoel CF.
Front Med (Lausanne). 2017;4 62
PMID: 28603714 DOI: 10.3389/fmed.2017.00062

PhenomeCentral: A Portal for Phenotypic and Genotypic Matchmaking of Patients with Rare Genetic Diseases. Orion J. Buske, Marta Girdea, Sergiu Dumitriu, Bailey Gallinger, Taila Hartley, Heather Trang, Andriy Misyura, Tal Friedman, Chandree Beaulieu, William P. Bone, Amanda E. Links, Nicole L. Washington, Melissa A. Haendel, Peter N. Robinson, Cornelius F. Boerkoel, David Adams, William A. Gahl, Kym M. Boycott, Michael Brudno.
Human Mutation 2015, 36(10):931-940
PMID: 26251998 DOI: 10.1002/humu.22851

Phenopolis: an open platform for harmonization and analysis of genetic and phenotypic data. Nikolas Pontikos, Jing Yu, Ismail Moghul, Lucy Withington, Fiona Blanco-Kelly, Tom Vulliamy, Tsz Lun, Ernest Wong, Cian Murphy, Valentina Cipriani, Alessia Fiorentino, Gavin Arno, Daniel Greene, Julius OB Jacobsen, Tristan Clark, David S. Gregory, Andrea M. Nemeth, Stephanie Halford, Chris F. Inglehearn, Susan Downes, Graeme C. Black, Andrew R. Webster, Alison J. Hardcastle, UKIRDC, Vincent Plagnol.
Bioinformatics 2017,33(15):2421-2423
PMID: 28334266 DOI:10.1093/bioinformatics/btx147

